### DISCUSSION TITLE
Support preCICE naively

### DISCUSSION THREAD

**Discussion by kursatyurt** [su2code/SU2#1972] [Ideas]
[preCICE](https://precice.org/) is a well-known partitioned multi-physics library with minimal invasive integration to existing codes.

There is an official[ SU2-preCICE adapter ](https://github.com/precice/su2-adapter) that exists, however, requires modification in the SU2 code base and recompilation. 

A recent development by @j-signorelli adapts the python wrappings of SU2 to implement adapters using python scripts [see](https://github.com/j-signorelli/su2-precice-pywrapper).

The advantage of using a python wrapper is that preCICE headers are not required during the compilation phase of the SU2. Only existing preCICE and python bindings are required if one wants to use preCICE.

In #1949 it has already been achieved that one can modify the existing solution on a marker.  Have access to the initial coordinates of the mesh (I am not sure about this). 

These futures are enough to reach an explicit coupling scheme. For implicit coupling, one needs to save the current state of the solution to memory (as IO is not efficient for a such task) and should be able to reload this solution.

I would like to get opinions about providing such functions in the official SU2 code.

**Comment by pcarruscag** [comment]
You save the current state into a numpy array with a couple lines of python. What else is needed?

**Reply by kursatyurt** [reply]
AFAIK access is possible with `driver.Solution()` at the current time. For a restart, the current time is not sufficient also `time_n` and `time_nM1` are required. From my understanding changing only solutions is not sufficient there is a need to change grid velocities and volumes are required. 

Is it possible to restart from the saved numpy arrays?

**Reply by pcarruscag** [reply]
What type of restart are we talking? SU2 loads that information itself when it restarts.

**Reply by kursatyurt** [reply]
When coupling with another solver implicitly, an in-memory alternative of restart provided by the `RESTART_SOL` option is required as those restart operations may happen many times and IO is not feasible.

some functions like [saveOldState](https://github.com/j-signorelli/su2-precice-pywrapper/blob/9cc69b9979eb4daaa04ae2ea6549d0e7262c3631/replacement_files/python_wrapper_structure.cpp#L543-L619) and [reloadOldState from memory](https://github.com/j-signorelli/su2-precice-pywrapper/blob/9cc69b9979eb4daaa04ae2ea6549d0e7262c3631/replacement_files/python_wrapper_structure.cpp#L301-L368)
 
Other function requirements for a preCICE adapter are mostly provided in the new python interface (access to the solution at boundary/initial mesh coordinates and setting mesh displacement).

**Reply by pcarruscag** [reply]
I don't understand why the coupling needs to modify the old solutions. Old solutions are for time domain problems, if that time-step passed the values are what they are. Maybe they are working around something else. At any rate, we are open to contributions to make SU2 easier to interface with other codes, as long as they are good solutions and integrated in a way that is compatible with most features of SU2. We have weekly developers meetings that can be used to discuss integration aspects.

**Reply by Nicola-Fonzi** [reply]
Hi @kursatyurt ,

Thanks for posting. What is the final goal you want to achieve? The fsi interface is meant to be flexible in terms of possible codes to couple with su2. Clearly, not as general as precice, but maybe you can already do what you are looking for

**Reply by kursatyurt** [reply]
@pcarruscag for explicit coupling schemes you do not need a restart or move backward in time but for implicit methods you need.

@Nicola-Fonzi I have already working setups with preCICE and adapters, migration to SU2 internal capability would be painful. The final goal is to provide the required interfaces for preCICE naively and sample scripts to users. Just requires a few additional methods and I don't want to patch those methods every time I upgrade SU2, providing them naively would be better.

I am not so familiar with the internal data structures of SU2, the two adapter codes I shared above have different approaches (they support different versions).

**Reply by Nicola-Fonzi** [reply]
I see, thank you for the explanation

**Reply by pcarruscag** [reply]
@kursatyurt they are calling driver.Update before the coupling for the current time step is finished, that is why they then need to restore the old state if the coupling is not converged yet.
Instead they should solve the coupling and only update (which pushes old solutions) once the coupling for the current time step is finished.
We have Implicit coupling in SU2 and that's how it works, no need to restore anything.

**Reply by kursatyurt** [reply]
@pcarruscag great, I will talk with preCICE developers if we can skip saving/reloading states (sounds no problem but to be sure).  if it internally works with preCICE saves some time and memory.

Out of context, I found it a bit confusing `Postprocess()` and `Postprocessing()` the first one does the post-processing of the solution, and the latter finalizes the driver right?

**Reply by pcarruscag** [reply]
That's correct, we can rename Postprocessing, I also think it is not very intuitive
